### PR TITLE
[KDA] Fix mixed exponent bases in Triton chunk prefill

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

KDA defines each per-channel decay gate in natural-log space. Let $g_t \in \mathbb{R}^{D_k}$ be the per-token log-decay gate and let its chunk-local cumulative gate be

$$
G_t = \sum_{s=1}^{t} g_s
$$

The decay that propagates information from position $j$ to position $i$ is

$$
D_{i,j} = \exp\left(G_i - G_j\right)
$$

image

The Triton chunk-prefill implementation mixed two exponent domains for the same cumulative-gate tensor:

- several intra-chunk kernels evaluated gate differences with `exp2`;
- the state-update, output, and fused-recompute paths evaluated them with natural `exp`;
- the cumulative sum remained in natural-log space.

Using `exp2` directly on a natural-log cumulative gate instead computes

$$
2^{G_i-G_j} = \exp\left(\ln(2)\left(G_i-G_j\right)\right)
$$

rather than $\exp(G_i-G_j)$. Since KDA gates are non-positive, multiplying the exponent by $\ln(2) < 1$ weakens the decay and retains too much history. Because different stages used different bases, the combined chunk computation did not correspond to one consistent KDA recurrence.

The correct conversion is performed once at the cumulative-gate producer:

$$
\widehat{G}_t = G_t \log_2(e) = \frac{G_t}{\ln(2)}
$$

All downstream KDA chunk kernels can then use base-2 exponentiation while preserving the original recurrence exactly:

$$
2^{\widehat{G}_i-\widehat{G}_j} = 2^{(G_i-G_j)/\ln(2)} = \exp\left(G_i-G_j\right)
$$

This affects the Triton KDA chunk-prefill path for both pre-activated gates and the fused raw-gate activation path. Recurrent decode continues to consume a raw per-token natural-log gate and is not affected. GDN also keeps its existing natural-exponent contract.

## Modifications

- Convert KDA's chunk-local cumulative gate from natural-log space to log2 space at its single producer boundary by passing `RCP_LN2` to both:
  - `kda_gate_chunk_cumsum`, for raw gates activated in the cumsum kernel;
  - `chunk_local_cumsum`, for gates already activated by the caller.
- Make all consumers of the KDA cumulative gate use `exp2`, including:
  - scaled KKT/QK construction;
  - standalone and fused W/U/KG recomputation;
  - intra-chunk fused solve/recompute;
  - recurrent-state chunk updates;
  - final chunk output construction.
- Add a compile-time `use_exp2` option to the shared `chunk_gated_delta_rule_fwd_h` state kernel. It defaults to `False`, preserving the existing GDN natural-exponent behavior; KDA passes `use_exp2=True`.
- Keep the XPU override's Python/kernel signature aligned with the shared state kernel.
- Add an independent pure-PyTorch recurrent oracle test. It checks both output and the in-place final state for:
  - fixed-length prefill crossing multiple 64-token chunks;
  - ragged varlen prefill around sub-chunk/chunk boundaries;
  - pre-activated gates and fused raw-gate activation.

This PR intentionally does not change recurrent decode, GDN's exponent domain, safe-gate kernel selection, or non-Triton KDA backends.

## Accuracy Tests

The new regression test uses a token-by-token PyTorch recurrence with natural `torch.exp` as an implementation-independent reference. Restoring the pre-fix sources makes the test fail by a large margin:

| Case | Pre-fix output relative RMSE | This PR |
| --- | ---: | --- |
| fixed length `[129]`, pre-activated gate | 19.7575% | output and final state < 1% |
| ragged `[15, 16, 17, 63, 65]`, fused raw gate | 14.9267% | output and final state < 1% |

- KDA gate cumsum and fused recurrent regression: `6 passed`.
- Shared GDN state-kernel regression: `1 passed`.
- Projected Triton KDA backend correctness: `1 passed`.

No end-to-end model accuracy benchmark was run for this kernel-only fix.

## Speed Tests and Profiling

No dedicated performance benchmark was run. The change adds no kernel launch, allocation, synchronization, or intermediate tensor:

- the `RCP_LN2` multiplication is fused into the existing cumsum kernel;
- natural `exp` instructions are replaced by `exp2` in KDA consumers;
- `USE_EXP2` is a Triton compile-time specialization, so it adds no runtime branch to either KDA or GDN.

The expected performance impact is neutral to slightly positive, but this PR makes no performance claim without a dedicated benchmark.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). No user-facing documentation change is required; the exponent-domain contract is documented in code and tests.
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). Kernel accuracy is covered above; a dedicated speed benchmark has not been run.
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`.
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #29887429616](https://github.com/sgl-project/sglang/actions/runs/29887429616)
Latest PR Test (Extra): :white_check_mark: [Run #29887429380](https://github.com/sgl-project/sglang/actions/runs/29887429380)

### CODE REVIEW

File: python/sglang/kernels/ops/attention/linear/kda_blackwell/__init__.py [hunk 1]
```diff
@@ -173,13 +173,18 @@ def chunk_kda_cutedsl(
...
+    )
 
     ones_beta = q.new_ones(1, T, Hv, dtype=torch.float32)
+    # The FLA kkt kernels consume log2-space gate cumsums (exp2-based); g_cu must
+    # stay natural-log for the cutedsl kernels below, so convert on a copy.
     M_kk, M_qk = chunk_kda_scaled_dot_kkt_fwd(
         q.unsqueeze(0).contiguous(),
         k.unsqueeze(0).contiguous(),
-        gk=g_cu.unsqueeze(0),
+        gk=(g_cu * RCP_LN2).unsqueeze(0),
```
Comment (kaixih) [inline python/sglang/kernels/ops/attention/linear/kda_blackwell/__init__.py]: do you think we should add a `gk_scale` so that we can apply it directly to the gate values when they are loaded in the Triton kernel, like
```
gk=g_cu.unsqueeze(0),
gk_scale=RCP_LN2,
```
This can avoid materializing a full-sized copy of the tensor.

File: test/registered/attention/test_kda_kernels.py [hunk 2]
```diff
@@ -245,6 +247,139 @@ def test_small_head_dim(self):
...
+    @torch.inference_mode()
+    def test_chunk_prefill_matches_natural_exp_recurrence(self):
+        device = get_device()
+        dtype = torch.bfloat16
+        num_heads, head_dim = 2, 64
+
+        cases = (
+            ([129], False, False),
+            ([15, 16, 17, 63, 65], True, True),
+        )
```
Comment (kaixih) [inline test/registered/attention/test_kda_kernels.py:299]: nit: do you think we should add one case to cover the non-fused path like:
```
([2] * 129, True, False),  # force non-fused path
```
I think this will lead to `_small_grid=False` since `B × num_chunks × H = 1 × 129 × 2 = 258 > 256`. Then, effectively, I guess both fuse_diagonal and fuse_recompute are disabled, forcing the standalone non-fused diagonal and recompute kernels. Right? We should cover this path.

File: python/sglang/kernels/ops/attention/linear/kda_blackwell/__init__.py (hunk 1 above)
Comment (yuan-luo) [reply python/sglang/kernels/ops/attention/linear/kda_blackwell/__init__.py]: Thanks. It's a good idea. I added a `gk_scale` param to `chunk_kda_scaled_dot_kkt_fwd`.

File: test/registered/attention/test_kda_kernels.py (hunk 2 above)
Comment (yuan-luo) [reply test/registered/attention/test_kda_kernels.py:299]: Agree. Added this case to cover the non-fused path.

**Review by kaixih — APPROVED** [review]

**Review by BBuf — APPROVED** [review]
